typed/remove_test.go

```diff
@@ -0,0 +1,457 @@
+package typed
+
+import (
+	"fmt"
+	"testing"
+
+	"sigs.k8s.io/structured-merge-diff/v3/fieldpath"
+	"sigs.k8s.io/structured-merge-diff/v3/value"
+)
+
+var (
+	// Short names for readable test cases.
+	_NS  = fieldpath.NewSet
+	_P   = fieldpath.MakePathOrDie
+	_KBF = fieldpath.KeyByFields
+	_V   = value.NewValueInterface
+	_I   = fieldpath.NewIndexPathElement
+)
+
+func TestRemoveDeduced(t *testing.T) {
+	var cases = []struct {
+		object          YAMLObject
+		remove          *fieldpath.Set
+		expect          YAMLObject
+		expectItemsOnly YAMLObject
+	}{
+		{
+			object:          `{}`,
+			remove:          _NS(_P("a")),
+			expect:          `{}`,
+			expectItemsOnly: `{}`,
+		},
+		{
+			object:          `{"a": "value"}`,
+			remove:          _NS(_P("a")),
+			expect:          `{}`,
+			expectItemsOnly: `{}`,
+		},
+		{
+			object:          `{"a": "value", "b": "value"}`,
+			remove:          _NS(_P("a")),
+			expect:          `{"b": "value"}`,
+			expectItemsOnly: `{"b": "value"}`,
+		},
+		{
+			object:          `{"a": "value", "b": {}}`,
+			remove:          _NS(_P("a")),
+			expect:          `{"b": {}}`,
+			expectItemsOnly: `{"b": {}}`,
+		},
+		{
+			object:          `{"a": "value", "b": {"c":"value"}}`,
+			remove:          _NS(_P("b")),
+			expect:          `{"a": "value"}`,
+			expectItemsOnly: `{"a": "value"}`,
+		},
+		{
+			object:          `{"a": "value", "b": {"c":"value"}}`,
+			remove:          _NS(_P("b", "c")),
+			expect:          `{"a": "value", "b": {}}`,
+			expectItemsOnly: `{"a": "value", "b": {}}`,
+		},
+		{
+			object:          `{"a": "value", "b": []}`,
+			remove:          _NS(_P("b")),
+			expect:          `{"a": "value"}`,
+			expectItemsOnly: `{"a": "value"}`,
+		},
+		{
+			object:          `{"a": "value", "b": ["c"]}`,
+			remove:          _NS(_P("b")),
+			expect:          `{"a": "value"}`,
+			expectItemsOnly: `{"a": "value"}`,
+		},
+		{
+			object: `{"a": "value", "b": ["c"]}`,
+			remove: _NS(_P("b", "c")),
+			// `c` won't get removed, as it is not a field, while being addressed as one in MakePathOrDie.
+			// There is a test for removing `c` below.
+			expect:          `{"a": "value", "b": ["c"]}`,
+			expectItemsOnly: `{"a": "value", "b": ["c"]}`,
+		},
+		{
+			object: `{"a": "value", "b": ["c", "d"]}`,
+			remove: _NS(_P("b", "c")),
+			// `c` won't get removed, as it is not a field, while being addressed as one in MakePathOrDie.
+			// There is a test for removing `c` below.
+			expect:          `{"a": "value", "b": ["c", "d"]}`,
+			expectItemsOnly: `{"a": "value", "b": ["c", "d"]}`,
+		},
+		{
+			object:          `{"a": "value", "b": ["c"]}`,
+			remove:          _NS(_P("b", _I(0))),
+			expect:          `{"a": "value", "b": []}`,
+			expectItemsOnly: `{"a": "value", "b": []}`,
+		},
+		{
+			object:          `{"a": "value", "b": ["c", "d"]}`,
+			remove:          _NS(_P("b", _I(1))),
```

Comment:
I don't see a test that removes by value, do we have one? Maybe below, but I suspect it belongs here